File: k9mail-library/src/main/java/com/fsck/k9/mail/message/MessageHeaderParser.java
package com.fsck.k9.mail.message;


import java.io.IOException;
import java.io.InputStream;

import com.fsck.k9.mail.MessagingException;
import com.fsck.k9.mail.Part;
import org.apache.james.mime4j.MimeException;
import org.apache.james.mime4j.parser.ContentHandler;
import org.apache.james.mime4j.parser.MimeStreamParser;
import org.apache.james.mime4j.stream.BodyDescriptor;
import org.apache.james.mime4j.stream.Field;
import org.apache.james.mime4j.stream.MimeConfig;


public class MessageHeaderParser {

    public static void parse(final Part part, InputStream headerInputStream) throws MessagingException {
        MimeStreamParser parser = getMimeStreamParser();
        parser.setContentHandler(new MessageHeaderParserContentHandler(part));

        try {
            parser.parse(headerInputStream);
        } catch (MimeException me) {
            throw new MessagingException("Error parsing headers", me);
        } catch (IOException e) {
            throw new MessagingException("I/O error parsing headers", e);
        }
    }

    private static MimeStreamParser getMimeStreamParser() {
        MimeConfig parserConfig = new MimeConfig();
        parserConfig.setMaxHeaderLen(-1);
        parserConfig.setMaxLineLen(-1);
        parserConfig.setMaxHeaderCount(-1);
        return new MimeStreamParser(parserConfig);
    }

    private static class MessageHeaderParserContentHandler implements ContentHandler {
        private final Part part;

        public MessageHeaderParserContentHandler(Part part) {
            this.part = part;
        }

        @Override
        public void field(Field rawField) throws MimeException {
            String name = rawField.getName();
            String raw = rawField.getRaw().toString();
            try {
                part.addRawHeader(name, raw);
            } catch (MessagingException e) {
                throw new RuntimeException(e);
            }
        }

        @Override
        public void startMessage() throws MimeException {
            /* do nothing */
        }

        @Override
        public void endMessage() throws MimeException {
            /* do nothing */
        }

        @Override
        public void startBodyPart() throws MimeException {
            /* do nothing */
        }

        @Override
        public void endBodyPart() throws MimeException {
            /* do nothing */
        }

        @Override
        public void startHeader() throws MimeException {
            /* do nothing */
        }

        @Override
        public void endHeader() throws MimeException {
            /* do nothing */
        }

        @Override
        public void preamble(InputStream is) throws MimeException, IOException {
            /* do nothing */
        }

        @Override
        public void epilogue(InputStream is) throws MimeException, IOException {
            /* do nothing */
        }

        @Override
        public void startMultipart(BodyDescriptor bd) throws MimeException {
            /* do nothing */
        }

        @Override
        public void endMultipart() throws MimeException {
            /* do nothing */
        }

        @Override
        public void body(BodyDescriptor bd, InputStream is) throws MimeException, IOException {
            /* do nothing */
        }

        @Override
        public void raw(InputStream is) throws MimeException, IOException {
            /* do nothing */
        }
    }
}

File: mail/common/src/main/java/com/fsck/k9/mail/message/MessageHeaderParser.kt
package com.fsck.k9.mail.message

import java.io.IOException
import java.io.InputStream
import net.thunderbird.core.common.exception.MessagingException
import org.apache.james.mime4j.MimeException
import org.apache.james.mime4j.parser.AbstractContentHandler
import org.apache.james.mime4j.parser.MimeStreamParser
import org.apache.james.mime4j.stream.Field
import org.apache.james.mime4j.stream.MimeConfig

object MessageHeaderParser {
    @Throws(MessagingException::class)
    @JvmStatic
    fun parse(headerInputStream: InputStream, collector: MessageHeaderCollector) {
        val parser = createMimeStreamParser().apply {
            setContentHandler(MessageHeaderParserContentHandler(collector))
        }

        try {
            parser.parse(headerInputStream)
        } catch (me: MimeException) {
            throw MessagingException("Error parsing headers", me)
        } catch (e: IOException) {
            throw MessagingException("I/O error parsing headers", e)
        }
    }

    private fun createMimeStreamParser(): MimeStreamParser {
        val parserConfig = MimeConfig.Builder()
            .setMaxHeaderLen(-1)
            .setMaxLineLen(-1)
            .setMaxHeaderCount(-1)
            .build()

        return MimeStreamParser(parserConfig)
    }

    private class MessageHeaderParserContentHandler(
        private val collector: MessageHeaderCollector,
    ) : AbstractContentHandler() {
        override fun field(rawField: Field) {
            val name = rawField.name
            val raw = rawField.raw.toString()
            collector.addRawHeader(name, raw)
        }
    }
}

fun interface MessageHeaderCollector {
    fun addRawHeader(name: String, raw: String)
}
